### PR TITLE
MVP

### PR DESCRIPTION
**Pull request by juanimedone**
# MVP
## Summary
This pull request evolves the project from scratch to a basic implementation of Conway's Game of Life. It includes the game logic, visual rendering, and some utilities for validation and testing.

### Game Logic Implementation
- Added the `Game` struct to represent the state of the game.
- Implemented the core game functions:
  - `new`: Initializes a new game with a **random** state.
  - `get_index`: Computes the index for a given cell position ((x,y) coordinates).
  - `count_neighbors`: Counts the number of live neighbors for a given cell.
  - `update`: Applies the rules of the game to update the grid.
  - `draw`: Renders the game state to the screen.

### Game Initialization and Rendering
- Created a `start` function to handle game initialization and rendering using the `macroquad` library.
- Added parameter validation in `validate_start_parameters` to ensure valid game settings.

### Testing
- Added unit tests for core game functions, including:
   - Validating start parameters.
   - Verifying index calculations.
   - Testing neighbor counting and game state updates.
- These include handling edge cases to ensure correct functionality.
- Fixed Clippy warnings and applied `cargo fmt` for code consistency and to prevent future bugs.

### CODE REVIEW

File: src/game.rs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    color::{BLACK, WHITE},
+    shapes::draw_rectangle,
+    window::{clear_background, next_frame},
+};
+
+/// Represents the state of the Game of Life.
+struct Game {
+    height: usize,
+    width: usize,
+    cells: Vec<bool>,
```
Comment (MarcosNicolau) [inline src/game.rs]: Why don't you use a matrix instead?

File: src/game.rs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - The x coordinate.
+    /// * `y` - The y coordinate.
+    ///
+    /// # Returns
+    ///
+    /// The index in the cells vector.
+    fn get_index(&self, x: i32, y: i32) -> usize {
```
Comment (MarcosNicolau) [inline src/game.rs]: You would avoid doing this calculations for example

File: src/game.rs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+    /// Draws the current game state.
+    ///
+    /// # Arguments
+    ///
+    /// * `cell_size` - The size of each cell.
+    fn draw(&self, cell_size: f32) {
+        for x in 0..self.width {
+            for y in 0..self.height {
+                let index = y * self.width + x;
```
Comment (MarcosNicolau) [inline src/game.rs]: shouldn't you use the `get_index` here?

File: src/game.rs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+/// # Arguments
+///
+/// * `height` - The height of the game grid.
+/// * `width` - The width of the game grid.
+/// * `cell_size` - The size of each cell.
+///
+/// # Returns
+///
+/// A message if an error ocurrs while validating the parameters, otherwise it loops infinitely.
+pub async fn start(height: usize, width: usize, cell_size: f32) -> Result<(), String> {
```
Comment (MarcosNicolau) [inline src/game.rs]: I would put the game loop inside the `Game` struct, since they are closely related and, indeed, it is the whole point of the game.

File: src/game.rs [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    /// # Arguments
+    ///
+    /// * `height` - The height of the game grid.
+    /// * `width` - The width of the game grid.
+    ///
+    /// # Returns
+    ///
+    /// A new `Game` instance.
+    fn new(height: usize, width: usize) -> Self {
+        let cells = (0..height * width).map(|_| rand::random()).collect();
```
Comment (MarcosNicolau) [inline src/game.rs]: aren't you creating a cell for every pixel here? shouldn't you divide the grid size by the cell size?

**Review by MarcosNicolau — CHANGES_REQUESTED** [review]
The game works fine, try to address the comments, merge it, and then add some new features, for example:

- Draw grid lines
- Add a start menu
- Ability to pause the game
- Ability to choose the initial alive seeds through a GUI
- Ability to change the game speed

File: src/game.rs (hunk 1 above)
Comment (vicentevieytes) [reply src/game.rs]: Agree, it'd be more readable with a Vec<Vec\<Bool> and indexing would be more intuitive.

**Review by vicentevieytes — CHANGES_REQUESTED** [review]
I think you should change the cells matrix from being a single array to an explicit matrix like Vec<Vec\<bool>. That you have to use get_index every time you want to index in the matrix seems unnecessarily complex to me.

The rest of the code seems correct to me, but i think you should make this change now before a lot more code is added to the project.

File: src/game.rs (hunk 5 above)
Comment (vicentevieytes) [reply src/game.rs]: This seems correct to me.

new just initializes the grid with the amount of elements indicated by height and width, then drawing is performed by the draw function that takes the cell size and multiplies it for every element in the grid.

Comment (MarcosNicolau) [reply src/game.rs]: Yeah sure, but what I meant is: aren't you creating more cells than needed?

Comment (MarcosNicolau) [reply src/game.rs]: For instance, if the screen is 1024x800 -> size(cells) = 819.200. Now if your cells size is, say, 5, then size(cells) = 1024/5 * 800/5 = 32.768

Comment (vicentevieytes) [reply src/game.rs]: I'm not sure honestly. Maybe if there are more cells than they fit in the screen there could be some issues with the way the program draws to the screen now that i think about it.

Comment (vicentevieytes) [reply src/game.rs]: > For instance, if the screen is 1024x800 -> size(cells) = 819.200. Now if your cells size is, say, 5, then size(cells) = 1024/5 * 800/5 = 32.768

The height and width passed to the `new` function are not necessarily the screen height and width. They are hard coded in main.rs in this case.

Comment (MarcosNicolau) [reply src/game.rs]: Imagine you pass the screen size as params, but it applies to any value

File: src/game.rs (hunk 1 above)
Comment (juanimedone) [reply src/game.rs]: You're right, I'll change that.

File: src/game.rs (hunk 5 above)
Comment (vicentevieytes) [reply src/game.rs]: Yeah, then it does get messed up i just tried it. But if you pass sreen-size/cell-size as params it works alright

File: src/game.rs (hunk 3 above)
Comment (juanimedone) [reply src/game.rs]: Yes, I changed `get_index`'s original code and forgot to update this part. Anyways, I'll change `cells` to be a `Vec<Vec<bool>>` so that indexing function won't be needed anymore.

File: src/game.rs (hunk 4 above)
Comment (juanimedone) [reply src/game.rs]: Ok, will do.

File: src/game.rs (hunk 5 above)
Comment (juanimedone) [reply src/game.rs]: Didn't notice that, I'll fix it.

**Comment by juanimedone** [comment]
I made several changes based on the comments you provided. These include:
- Turning `cells` to a matrix (`Vec<Vec<bool>>`) [like you proposed](https://github.com/juanimedone/conways-game-of-life/pull/1#discussion_r1684944107) in order to get more understandability and to avoid unnecessary calculations.
- Moving the game's main loop (`start` function) inside `Game` as a struct method. Refers to this [comment](https://github.com/juanimedone/conways-game-of-life/pull/1#discussion_r1684946951).
- Fixing the [grid size issue](https://github.com/juanimedone/conways-game-of-life/pull/1#discussion_r1684950525).

Please check them out and if everything looks fine for you I would suggest merging this PR and start working on the next features.

**Review by vicentevieytes — APPROVED** [review]
🚀💫Looks good

**Review by MarcosNicolau — APPROVED** [review]
LGTM 👍
